File: android-kotlin/QuickStartTasks/app/src/main/java/live/ditto/quickstart/tasks/data/DittoManager.kt
package live.ditto.quickstart.tasks.data

import android.util.Log
import androidx.activity.result.launch
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import live.ditto.Ditto
import live.ditto.DittoError
import live.ditto.DittoIdentity
import live.ditto.DittoLogLevel
import live.ditto.DittoLogger
import live.ditto.DittoStoreObserver
import live.ditto.DittoSyncSubscription
import live.ditto.android.AndroidDittoDependencies
import live.ditto.quickstart.tasks.models.DittoConfig
import live.ditto.quickstart.tasks.services.ErrorService

class DittoManager(
    override val dittoConfig: DittoConfig,
    androidDittoDependencies: AndroidDittoDependencies,
    private val errorService: ErrorService
) : DataManager {
    companion object {
        private const val TAG = "DittoManager"
    }

    // Use Dispatchers.IO for the scope if most operations will be I/O bound
    //https://developer.android.com/kotlin/coroutines/coroutines-adv#main-safety
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    var ditto: Ditto? = null
    private var subscription: DittoSyncSubscription? = null
    private var storeObserver: DittoStoreObserver? = null

    init {
        try {
            DittoLogger.minimumLogLevel = DittoLogLevel.DEBUG

            val identity = DittoIdentity.OnlinePlayground(
                androidDittoDependencies,
                dittoConfig.appId,
                dittoConfig.authToken,
                false,
                dittoConfig.authUrl
            )
            this.ditto = Ditto(androidDittoDependencies, identity)

            ditto?.let { ditto ->
                scope.launch {
                    try {
                        // Set the Ditto Websocket URL
                        ditto.updateTransportConfig { config ->
                            config.connect.websocketUrls.add(dittoConfig.websocketUrl)
                        }

                        // disable sync with v3 peers, required for DQL to work
                        ditto.disableSyncWithV3()

                        // Disable DQL strict mode
                        // when set to false, collection definitions are no longer required. SELECT queries will return and display all fields by default.
                        // https://docs.ditto.live/dql/strict-mode
                        ditto.store.execute("ALTER SYSTEM SET DQL_STRICT_MODE = false")
                    } catch (e: Exception) {
                        errorService.showError("Error during initial database operation: ${e.message}")
                        Log.e(TAG, "Error during initial database operation", e)
                    }
                }
            }
        } catch (e: DittoError) {
            errorService.showError("Failed to initialize Ditto: ${e.message}")
        }
    }

    /**
     * Populates the Ditto tasks collection with initial seed data if it's empty.
     *
     * This method:
     * - Creates a set of predefined tasks with unique IDs and titles
     * - Inserts each task into the Ditto store using DQL (Ditto Query Language)
     * - Uses the INITIAL keyword to only insert if documents don't already exist
     *
     * The initial tasks include common todo items like:
     * - Buy groceries
     * - Clean the kitchen
     * - Schedule dentist appointment
     * - Pay bills
     *
     * Implementation details:
     * - Uses DQL INSERT statement with INITIAL keyword to prevent duplicates
     * - Each task document contains _id, title, done, and deleted fields
     * - Handles errors gracefully and logs them for debugging
     *
     * @throws Exception if the insert operations fail
     * Ditto Docs:
     * https://docs.ditto.live/sdk/latest/crud/write#inserting-documents
     */
    override suspend fun populateTaskCollection() {
        return withContext(Dispatchers.IO) {
            val tasks = listOf(
                TaskModel("50191411-4C46-4940-8B72-5F8017A04FA7", "Buy groceries"),
                TaskModel("6DA283DA-8CFE-4526-A6FA-D385089364E5", "Clean the kitchen"),
                TaskModel("5303DDF8-0E72-4FEB-9E82-4B007E5797F0", "Schedule dentist appointment"),
                TaskModel("38411F1B-6B49-4346-90C3-0B16CE97E174", "Pay bills")
            )

            tasks.forEach { task ->
                try {
                    // https://docs.ditto.live/sdk/latest/crud/write#inserting-documents
                    ditto?.store?.execute(
                        "INSERT INTO tasks INITIAL DOCUMENTS (:task)",
                        mapOf(
                            "task" to mapOf(
                                "_id" to task._id,
                                "title" to task.title,
                                "done" to task.done,
                                "deleted" to task.deleted,
                            )
                        )
                    )
                } catch (e: Exception) {
                    errorService.showError("Unable to insert initial documents: ${e.message}")
                    Log.e(TAG, "Unable to insert initial document", e)
                }
            }
        }
    }

    /**
     * Creates a Flow that observes and emits changes to the tasks collection in Ditto.
     * https://docs.ditto.live/sdk/latest/crud/read#using-args-to-query-dynamic-values
     *
     * This method:
     * - Sets up a live query observer for the tasks collection
     * - Emits a new list of TaskModel objects whenever the data changes
     * - Automatically cleans up the observer when the flow is cancelled
     *
     * The query:
     * - Returns all non-archived TaskModel (done = false)
     * - Transforms raw Ditto data into TaskModel objects
     *
     * @return Flow<List<TaskModel>> A flow that emits updated lists of Tasks
     * whenever changes occur in the collection
     *
     */
    override fun getTaskModels(): Flow<List<TaskModel>> = callbackFlow {
        try {
            val query = "SELECT * FROM tasks WHERE NOT deleted"

            storeObserver = ditto?.store?.registerObserver(query) { results ->
                val items = results.items.map { item ->
                    TaskModel.fromMap(item.value)
                }
                trySend(items)
            }
            awaitClose {
                storeObserver?.close()
                storeObserver = null
            }
        } catch (e: Exception) {
            errorService.showError("Failed to setup observer for getting taskModels: ${e.message}")
            Log.e(TAG, "Failed to setup observer for getting taskModels", e)
        }
    }

    /**
     * Enables or disables synchronization with the Ditto cloud (BigPeer).
     *
     * This method:
     * - Manages the sync state of the Ditto instance
     * - Persists the sync preference in DataStore
     * - Updates the UI state through LiveData
     *
     * Implementation details:
     * - Retrieves default sync state from preferences if not specified
     * - Starts sync if enabled and not already active
     * - Stops sync if disabled and currently active
     * - Updates both DataStore and LiveData to maintain state
     *
     * @param enabled Boolean value to set sync state to. If null, uses stored preference
     *
     * @see startSync()
     * @see stopSync()
     */
    override suspend fun setSyncEnabled(enabled: Boolean) {
        //get default value from preferences if not set
        if (enabled  && !(ditto?.isSyncActive)!!) {
            startSync()
        } else {
            stopSync()
        }
    }

    /**
     * Starts the Ditto sync process and sets up the initial subscription.
     *
     * This method:
     * - Initiates the sync process
     * - Creates a subscription to the tasks collection
     * - Ensures real-time synchronization of task data across devices
     *
     * The subscription:
     * - Queries all documents in the tasks collection
     * - Maintains the subscription until sync is stopped or the app is terminated
     * - Automatically syncs changes between the local store and the Ditto cloud
     *
     * Implementation details:
     * - Uses Ditto's sync API to establish connection with the BigPeer
     * - Sets up a subscription using DQL to monitor the tasks collection
     * - Handles errors gracefully and logs them for debugging
     *
     * @throws Exception if sync initialization or subscription setup fails
     *
     * Ditto Docs:
     * https://docs.ditto.live/sdk/latest/sync/start-and-stop-sync
     * https://docs.ditto.live/sdk/latest/sync/syncing-data
     *
     */
    private suspend fun startSync() {
        return withContext(Dispatchers.IO) {
            try {
                ditto?.let {
                    it.startSync()

                    val subscriptionQuery = "SELECT * from tasks"
                    subscription = it.sync.registerSubscription(subscriptionQuery)
                }
            } catch (e: Exception) {
                errorService.showError("Failed to start ditto sync: ${e.message}")
                Log.e(TAG, "Failed to start ditto sync", e)
            }
        }
    }

    /**
     * Stops the Ditto sync process and cleans up associated resources.
     *
     * This method:
     * - Cancels any active subscriptions to the tasks collection
     * - Stops the sync process with the Ditto cloud (BigPeer)
     * - Cleans up resources to prevent memory leaks
     *
     * Implementation details:
     * - Closes the subscription and sets it to null
     * - Calls Ditto's stopSync() to terminate the sync process
     * - Should be called when sync is no longer needed or app is terminating
     *
     * @see Ditto.stopSync()
     * @see DittoSyncSubscription.close()
     */
    suspend fun stopSync() {
        return withContext(Dispatchers.IO) {
            //implements stopping sync by calling cancel and setting object to nil
            //and then running stopSync function
            // https://docs.ditto.live/sdk/latest/sync/syncing-data#canceling-subscriptions
            subscription?.close()
            subscription = null
            ditto?.stopSync()
        }
    }



    /**
     * Creates a new TaskModel document in the Ditto store.
     * https://docs.ditto.live/sdk/latest/crud/create#creating-documents
     *
     * Implementation details:
     * - Inserts a complete document with all required fields using DQL
     *
     * @param taskModel The new TaskModel object to insert
     */
    override suspend fun insertTaskModel(taskModel: TaskModel) {
        return withContext(Dispatchers.IO) {
            try {
                val query = "INSERT INTO tasks DOCUMENTS (:newTask)"
                ditto?.store?.execute(
                    query,
                    mapOf(
                        "newTask" to mapOf(
                            "_id" to taskModel._id,
                            "title" to taskModel.title,
                            "done" to taskModel.done,
                            "deleted" to taskModel.deleted
                        )
                    )
                )
            } catch (e: Exception) {
                errorService.showError("Failed to add taskModel: ${e.message}")
                Log.e(TAG, "Failed to add taskModel:", e)
            }
        }
    }

    /**
     * Updates existing TaskModel's properties in the Ditto store.
     * https://docs.ditto.live/sdk/latest/crud/update#updating
     * Implementation details:
     * - Uses DQL to update the document
     *
     * @param taskModel The TaskModel object containing updated values
     */
    override suspend fun updateTaskModel(taskModel: TaskModel) {
        return withContext(Dispatchers.IO) {
            try {
                // https://docs.ditto.live/dql/update#basic-update
                // https://docs.ditto.live/sdk/latest/crud/update#updating
                val query = """
                UPDATE tasks
                SET title = :title,
                    done = :done,
                    deleted = :deleted
                WHERE _id = :_id 
                """

                // https://docs.ditto.live/sdk/latest/crud/update#updating
                ditto?.store?.execute(
                    query,
                    mapOf(
                        "title" to taskModel.title,
                        "done" to taskModel.done,
                        "deleted" to taskModel.deleted,
                        "_id" to taskModel._id
                    )
                )
            } catch (e: Exception) {
                errorService.showError("Failed to update taskModel: ${e.message}")
                Log.e(TAG, "Failed to update taskModel:", e)
            }
        }
    }

    /**
     * Updates an existing TaskModel in the Ditto store, setting the done
     * field to true.
     *
     * This method:
     * - Updates the done field to true
     * - Maintains the taskModel's ID and references
     * - Triggers a sync with other devices
     *
     * The taskModel object should have:
     * - A unique ID
     * - All required fields populated
     *
     * @param id The _id TaskModel object to update
     * @throws Exception if the insert operation fails
     */
    override suspend fun toggleComplete(id: String) {
        return withContext(Dispatchers.IO) {
            try {
                ditto?.let {
                    val doc = it.store.execute(
                        "SELECT * FROM tasks WHERE _id = :_id AND NOT deleted",
                        mapOf("_id" to id)
                    ).items.first()

                    val done = doc.value["done"] as Boolean

                    // https://docs.ditto.live/dql/update#basic-update
                    // https://docs.ditto.live/sdk/latest/crud/update#updating
                    val query = """
                        UPDATE tasks
                        SET done = :done 
                        WHERE _id == :_id
                    """

                    // https://docs.ditto.live/sdk/latest/crud/update#updating
                    it.store.execute(
                        query,
                        mapOf(
                            "done" to !done,
                            "_id" to id
                        )
                    )
                }
            } catch (e: Exception) {
                errorService.showError("Failed to update taskModel: ${e.message}")
                Log.e(TAG, "Failed to update taskModel:", e)
            }
        }
    }

    /**
     * Marks a TaskModel as archived in the Ditto store.
     * This uses the 'Soft-Delete' pattern:  https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
     *
     * Implementation details:
     * - Uses a simple DQL UPDATE query to set done flag
     *
     * @param id The _id of the TaskModel to archive
     */
    override suspend fun deleteTaskModel(id: String) {
        return withContext(Dispatchers.IO) {
            try {
                // https://docs.ditto.live/sdk/latest/crud/delete#soft-delete-pattern
                val query = """
                UPDATE tasks
                SET deleted = true
                WHERE _id = :_id 
                """
                ditto?.store?.execute(
                    query,
                    mapOf(
                        "_id" to id,
                    )
                )
            } catch (e: Exception) {
                errorService.showError("Failed to archive taskModel: ${e.message}")
                Log.e(TAG, "Failed to archive taskModel:", e)
            }
        }
    }
}

File: dotnet-maui/DittoMauiTasksApp/Data/DittoManager.cs
using DittoSDK;
using Microsoft.Extensions.Logging;
using System.Text.Json;
using System.Collections.ObjectModel;
using DittoMauiTasksApp.Model;

namespace DittoMauiTasksApp.Data;

/// <summary>
/// Manages Ditto database operations for the tasks application, including CRUD operations,
/// data synchronization, and real-time data observation. This class serves as the primary
/// data access layer for task management functionality.
/// </summary>

public class DittoManager(ILogger<DittoManager> logger)
    : IDataManager
{
    private readonly ILogger<DittoManager> _logger = logger;
    private const string SelectQuery = "SELECT * FROM tasks WHERE NOT deleted";

#nullable enable
    private Ditto? _ditto = null;
    private DittoStoreObserver? _dittoStoreObserver = null;
    private DittoSyncSubscription? _syncSubscription = null;

    /// <summary>
    /// DittoDiffer instance used to calculate the delta changes between syncs.
    /// This enables efficient UI updates by only processing changed items.
    /// </summary>
    /// <see href="https://docs.ditto.live/sdk/latest/crud/read#diffing-results"/>
    private readonly DittoDiffer _dittoDiffer = new DittoDiffer();
    
    /// <summary>
    /// Cache of the previous query results for diffing operations.
    /// </summary>
    private List<DittoQueryResultItem> _previousItems = [];

    /// <summary>
    /// Initializes the Ditto instance with the specified configuration and data path.
    /// Sets up the Ditto identity, transport configuration, and initializes the database
    /// with default tasks if it's empty.
    /// </summary>
    /// <param name="dittoConfig">The Ditto configuration containing app ID, tokens, and URLs.</param>
    /// <param name="dataPath">The file system path where Ditto should store its data.</param>
    /// <returns>A task that represents the asynchronous initialization operation.</returns>
    /// <exception cref="Exception">Thrown when Ditto initialization fails.</exception>
    public async Task Initialize(DittoConfig dittoConfig, string dataPath)
    {
        try
        {
            _ditto = new Ditto(DittoIdentity
                .OnlinePlayground(
                    dittoConfig.AppId,
                    dittoConfig.PlaygroundToken,
                    false, // This is required to be set to false, to use the correct URLs with Ditto Portal
                    dittoConfig.AuthUrl), dataPath);

            _ditto.UpdateTransportConfig(config =>
            {
                // Add the websocket URL to the transport configuration.
                config.Connect.WebsocketUrls.Add(dittoConfig.WebsocketUrl);
            });

            // disable sync with v3 peers, required for DQL
            _ditto.DisableSyncWithV3();

            await _ditto.Store.ExecuteAsync("ALTER SYSTEM SET DQL_STRICT_MODE = false");
            await InsertInitialTasks();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, ex.Message);
        }    
           
    }

    /// <summary>
    /// Initializes the Ditto instance with the specified configuration using the default
    /// local application data path for data storage.
    /// </summary>
    /// <param name="dittoConfig">The Ditto configuration containing app ID, tokens, and URLs.</param>
    /// <returns>A task that represents the asynchronous initialization operation.</returns>
    public async Task Initialize(DittoConfig dittoConfig)
    {
        var dataPath = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "ditto");
        await Initialize(dittoConfig, dataPath);
    }

    /// <summary>
    /// Registers a store observer to monitor changes in the tasks collection and automatically
    /// update the provided ObservableCollection. Uses diffing to efficiently update only
    /// changed items, preventing unnecessary UI repaints.
    /// </summary>
    /// <param name="tasks">The ObservableCollection to be updated when task data changes.</param>
    /// <see href="https://docs.ditto.live/sdk/latest/crud/observing-data-changes#setting-up-store-observers"/>
    /// <see href="https://docs.ditto.live/sdk/latest/crud/read#diffing-results"/>
    public void RegisterObservers(ObservableCollection<DittoTask> tasks)
    {
        // Register observer, which runs against the local database on this peer
        // https://docs.ditto.live/sdk/latest/crud/observing-data-changes#setting-up-store-observers
        _dittoStoreObserver = _ditto?.Store.RegisterObserver(SelectQuery, (queryResult) =>
        {
            try
            {
                // .NET specifically with ObservableCollection will repaint the entire list if you just clear it and add new items.
                // Diffing the items allows us to only update the changed items, but at the expense of some memory and CPU usage.
                // This is a trade-off that is acceptable for this example, but may lead to problems on larger datasets.
                // https://docs.ditto.live/sdk/latest/crud/read#diffing-results
                var diff = _dittoDiffer.Diff(queryResult.Items);

                // For testing, we'll execute directly since we're not in a UI context
                try
                {
                    //handle deletions
                    foreach (var index in diff.Deletions)
                    {
                        var item = _previousItems[index];
                        var taskItem = JsonSerializer
                            .Deserialize<DittoTask>(_previousItems[index].JsonString());
                        if (taskItem != null)
                        {
                            tasks.Remove(tasks.First(x => x.Id == taskItem.Id));
                        }
                    }

                    //handle insertions
                    foreach (var index in diff.Insertions)
                    {
                        var item = queryResult.Items[index];
                        var taskItem = JsonSerializer
                            .Deserialize<DittoTask>(item.JsonString());
                        if (taskItem != null)
                        {
                            tasks.Add(taskItem);
                        }
                    }

                    //handle updates
                    foreach (var index in diff.Updates)
                    {
                        var item = queryResult.Items[index];
                        var taskItem = JsonSerializer
                            .Deserialize<DittoTask>(item.JsonString());
                        if (taskItem != null)
                        {
                            var existingTask = tasks.FirstOrDefault(x => x.Id == taskItem.Id);
                            if (existingTask != null)
                            {
                                //implement soft delete for the deleted field
                                if (taskItem.Deleted)
                                {
                                    tasks.Remove(existingTask);
                                }
                                else
                                {
                                    existingTask.Title = taskItem.Title;
                                    existingTask.Done = taskItem.Done;
                                }
                            }
                        }
                    }
                }
                catch (Exception e)
                {
                    _logger.LogError($"TasksPageViewModel: Error: Unable to update list view model: {e.Message}");
                }

                _previousItems = queryResult.Items;
            }
            catch (Exception e)
            {
                _logger.LogError($"TasksPageViewModel: Error: Unable to process tasks collection change: {e.Message}");
            }
        });
    }

    /// <summary>
    /// Starts the Ditto synchronization process and registers a subscription for the tasks collection.
    /// This enables real-time data synchronization with other peers and the Ditto cloud.
    /// </summary>
    /// <see href="https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions"/>
    public void StartSync()
    {
        _ditto?.StartSync();

        // Register a subscription, which determines what data syncs to this peer
        // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
        _syncSubscription = _ditto?.Sync.RegisterSubscription(SelectQuery);
    }

    /// <summary>
    /// Stops the Ditto synchronization process and cancels the active subscription.
    /// This should be called when sync is no longer needed or when the application is shutting down.
    /// </summary>
    public void StopSync()
    {
        _ditto?.StopSync();
        _syncSubscription?.Cancel();
    }

    /// <summary>
    /// Populates the tasks collection with initial seed data if it's empty.
    /// Inserts four predefined tasks using the INITIAL keyword to prevent duplicates.
    /// </summary>
    /// <returns>A task that represents the asynchronous insertion operation.</returns>
    /// <exception cref="Exception">Thrown when the insertion operation fails.</exception>
     public async Task InsertInitialTasks()
     {
        try
        {
            var initialTasks = new List<Dictionary<string, object>>
            {
                new()
                {
                    { "_id", "50191411-4C46-4940-8B72-5F8017A04FA7" },
                    { "title", "Buy groceries" },
                    { "done", false },
                    { "deleted", false }
                },
                new()
                {
                    { "_id", "6DA283DA-8CFE-4526-A6FA-D385089364E5" },
                    { "title", "Clean the kitchen" },
                    { "done", false },
                    { "deleted", false }
                },
                new()
                {
                    { "_id", "5303DDF8-0E72-4FEB-9E82-4B007E5797F0" },
                    { "title", "Schedule dentist appointment" },
                    { "done", false },
                    { "deleted", false }
                },
                new()
                {
                    { "_id", "38411F1B-6B49-4346-90C3-0B16CE97E174" },
                    { "title", "Pay bills" },
                    { "done", false },
                    { "deleted", false }
                }
            };

            const string insertCommand = "INSERT INTO tasks INITIAL DOCUMENTS (:task)";

            if (_ditto != null)
            {
                foreach (var task in initialTasks)
                {
                    await _ditto.Store.ExecuteAsync(insertCommand, new Dictionary<string, object>()
                    {
                        { "task", task }
                    });
                }
            }
        }
        catch (Exception e)
        {
            _logger.LogError($"TasksPageViewModel: Error adding initial tasks: {e.Message}");
        }
    }

    /// <summary>
    /// Adds a new task to the tasks collection in the Ditto store.
    /// </summary>
    /// <param name="document">A dictionary containing the task data with keys: _id, title, done, deleted.</param>
    /// <returns>A task that represents the asynchronous insertion operation.</returns>
    /// <exception cref="Exception">Thrown when the insertion operation fails.</exception>
    public async Task AddTask(Dictionary<string, object> document)
    {
        const string insertCommand = "INSERT INTO tasks DOCUMENTS (:doc)";
        if (_ditto != null)
        {
            await _ditto.Store.ExecuteAsync(insertCommand, new Dictionary<string, object>()
            {
                { "doc", document }
            });
        }
    }

    /// <summary>
    /// Updates the title of an existing task in the Ditto store.
    /// </summary>
    /// <param name="task">The DittoTask object containing the updated title and task ID.</param>
    /// <returns>A task that represents the asynchronous update operation.</returns>
    /// <exception cref="Exception">Thrown when the update operation fails.</exception>
    public async Task EditTask(DittoTask task)
    {
        if (_ditto != null)
        {
            const string updateQuery = "UPDATE tasks SET title = :title WHERE _id = :id";
            await _ditto.Store.ExecuteAsync(updateQuery, new Dictionary<string, object>()
            {
                { "title", task.Title },
                { "id", task.Id }
            });
        }
    }

    /// <summary>
    /// Marks a task as deleted using the soft delete pattern. The task remains in the database
    /// but is marked as deleted and will not appear in normal queries.
    /// </summary>
    /// <param name="task">The DittoTask object to be marked as deleted.</param>
    /// <returns>A task that represents the asynchronous deletion operation.</returns>
    /// <exception cref="Exception">Thrown when the deletion operation fails.</exception>
    public async Task DeleteTask(DittoTask task)
    {
        if (_ditto != null)
        {
            const string updateQuery = "UPDATE tasks SET deleted = true WHERE _id = :id";
            await _ditto.Store.ExecuteAsync(updateQuery, new Dictionary<string, object>()
            {
                { "id", task.Id }
            });
        }
    }

    /// <summary>
    /// Updates the completion status of a task in the Ditto store. Only updates if the
    /// current done state differs from the new state to avoid unnecessary operations.
    /// </summary>
    /// <param name="task">The DittoTask object containing the updated done status and task ID.</param>
    /// <returns>A task that represents the asynchronous update operation.</returns>
    /// <exception cref="Exception">Thrown when the update operation fails.</exception>
    public async Task UpdateTaskDoneAsync(DittoTask task)
    {
        if (_ditto != null)
        {
            var doc = new Dictionary<string, object>
            {
                { "newDoneState", task.Done },
                { "id", task.Id }
            };

            const string updateQuery =
                "UPDATE tasks SET done = :newDoneState WHERE _id = :id AND done != :newDoneState";
            await _ditto.Store.ExecuteAsync(updateQuery, doc);
        }
    }
}